raft: send MsgAppResp to latest leader after handling snapshot (#140233)
cockroachdb/cockroach: pkg/raft/testdata/snapshotOld_and_new_leader.txt

# This test sets up a 3-node Raft cluster to verify redundant snapshot handling
# across leadership changes. It ensures that a slow follower (Node 3) correctly
# processes snapshots from both the previous (Node 1) and current (Node 2)
# leader.  And also the new leader (Node 2) can avoid wasting necessary time in
# StateSnapshot.

# In the case that this slow follower happens to become the leaseholder,
# it is crucial for the leader node's progress tracker to have the leaseholder
# node transition from StateSnapshot -> StateReplicate ASAP

# Summary of steps:
# 1. Node 1 (previous leader) sends a snapshot to Node 3 (slow follower).
# 2. Leadership changes from Node 1 → Node 2.
# 3. Node 2 starts sending MsgApp to all followers, including Node 3.
# 4. Node 3, now aware of the leader change, responds with MsgAppResp to Node 2.

> #    A hint index is included in MsgAppResp to indicate node 3's progress

Reviewer: The hint index is unrelated here. Can drop this line.

> # 5. If Node 2 has already compacted past Node 3's hint index X, it sends a

Reviewer: ```suggestion
# 5. Node 2 has already compacted past Node 3's match index, so it sends a
```

#    snapshot to Node 3.
# 6. Node 3 receives the snapshot from previous term leader first.
#    It sends MsgAppResp to both Node 1 and Node 2.
# 7. Node 2 had previously marked Node 3 as StateSnapshot when sending its

> #    own snapshot 2->3. If Node 3 responds to Snapshot 1->3, Node 2 should revert
> #    Node 3 back to StateReplicate.

Reviewer: ```suggestion
#    own snapshot 2->3. If Node 3 responds to Snapshot 1->3, Node 2 transitions
#    Node 3 to StateReplicate.
```

# from StateSnapshot
# 2 then sends MsgApp 2->3 for entry 16, since 3 is caught up and in state
# replicate from the perspective of 2
stabilize 2
----
> 2 receiving messages
  3->2 MsgAppResp Term:2 Log:1/15 Rejected (Hint: 11) Commit:11
  DEBUG 2 received MsgAppResp(rejected, hint: (index 11, term 1)) from 3 for index 15
  DEBUG 2 decreased progress of 3 to [StateProbe match=0 next=12 sentCommit=11 matchCommit=11]
  DEBUG 2 [firstindex: 16, commit: 16] sent snapshot[index: 16, term: 2] to 3 [StateProbe match=0 next=12 sentCommit=11 matchCommit=11]
  DEBUG 2 paused sending replication messages to 3 [StateSnapshot match=0 next=17 sentCommit=16 matchCommit=11 paused pendingSnap=16]
  3->2 MsgVoteResp Term:2 Log:0/0 Rejected (Hint: 0)
  3->2 MsgFortifyLeaderResp Term:2 Log:0/0 LeadEpoch:1
> 2 handling Ready
  Ready MustSync=false:
  Messages:
  2->3 MsgSnap Term:2 Log:0/0
    Snapshot: Index:16 Term:2 ConfState:Voters:[1 2 3] VotersOutgoing:[] Learners:[] LearnersNext:[] AutoLeave:false

status 2
----
1: StateReplicate match=16 next=17 sentCommit=16 matchCommit=16
2: StateReplicate match=16 next=17 sentCommit=15 matchCommit=15

> 3: StateSnapshot match=0 next=17 sentCommit=16 matchCommit=11 paused pendingSnap=16

Reviewer: I think we want to observe exit from `StateSnapshot`, but we see we're still in `StateSnapshot` here.

  Snapshot Index:15 Term:1 ConfState:Voters:[1 2 3] VotersOutgoing:[] Learners:[] LearnersNext:[] AutoLeave:false
  Messages:
  3->2 MsgAppResp Term:2 Log:1/15 Rejected (Hint: 11) Commit:11

>   3->2 MsgAppResp Term:2 Log:0/15 Commit:15

Author: The slow follower Node 3 sends MsgAppResp to the current leader after applying the snapshot from previous leader.

  DEBUG 2 decreased progress of 3 to [StateProbe match=0 next=12 sentCommit=11 matchCommit=11]
  DEBUG 2 [firstindex: 16, commit: 16] sent snapshot[index: 16, term: 2] to 3 [StateProbe match=0 next=12 sentCommit=11 matchCommit=11]
  DEBUG 2 paused sending replication messages to 3 [StateSnapshot match=0 next=17 sentCommit=16 matchCommit=11 paused pendingSnap=16]

>   3->2 MsgAppResp Term:2 Log:0/15 Commit:15

Author: Current leader Node 2 sees Node 3 is caught up to commit index 15, which is higher than Node 2's firstIndex (not compacted)
(NB: although this message does not match a message Node 2 sent, as long as it indicates Node 3 is caught up, Node 2 can update its progress tracker for Node 3)

>   DEBUG 2 recovered from needing snapshot, resumed sending replication messages to 3 [StateSnapshot match=15 next=17 sentCommit=16 matchCommit=15 paused pendingSnap=16]

Author: Current leader Node 2 sees Node 3 is caught up to commit index 15, which is higher than Node 2's firstIndex (not compacted)
(NB: although this message does not match a message Node 2 sent, as long as it indicates Node 3 is caught up, Node 2 can update its progress tracker for Node 3)

> 2 handling Ready
  Ready MustSync=false:
  Messages:
  2->3 MsgSnap Term:2 Log:0/0
    Snapshot: Index:16 Term:2 ConfState:Voters:[1 2 3] VotersOutgoing:[] Learners:[] LearnersNext:[] AutoLeave:false

>   2->3 MsgApp Term:2 Log:1/15 Commit:16 Entries:[2/16 EntryNormal ""]

Author: Node 2 starts sending normal MsgApp in replicate mode to Node3. Instead of waiting for the response of its own snapshot to Node3.

(NB: its ok if this `2->3 MsgApp Term:2 Log:1/15 Commit:16 Entries:[2/16 EntryNormal ""` is stale when Node 3 process it. It will just be ignored.)
